ngraph/test/runtime/interpreter/evaluates_map.cpp

```diff
@@ -0,0 +1,1187 @@
+#include "reference/selu.hpp"
+
+using namespace ngraph;
+using namespace std;
+
+namespace
+{
+    template <element::Type_t ET>
+    bool evaluate(shared_ptr<Node> op,
+                  const HostTensorVector& outputs,
+                  const HostTensorVector& inputs)
+    {
+        return false;
+    }
+
+    template <element::Type_t ET>
+    bool evaluate(const shared_ptr<op::v1::Convolution>& op,
+                  const HostTensorVector& outputs,
+                  const HostTensorVector& inputs)
+    {
+        const auto filter_data = inputs[1]->get_data_ptr<ET>();
+        auto out_data_ptr = outputs[0]->get_data_ptr<ET>();
+        const auto in_data_ptr = inputs[0]->get_data_ptr<ET>();
+        const auto& out_shape = outputs[0]->get_shape();
+        const auto& in_shape = inputs[0]->get_shape();
+        const auto& filter_shape = inputs[1]->get_shape();
+        Strides in_dilation(std::vector<size_t>(in_shape.size() - 2));
+        std::fill(in_dilation.begin(), in_dilation.end(), 1);
+        runtime::reference::convolution<typename element_type_traits<ET>::value_type>(
+            in_data_ptr,
+            filter_data,
+            out_data_ptr,
+            in_shape,
+            filter_shape,
+            out_shape,
+            op->get_strides(),
+            op->get_dilations(),
+            op->get_pads_begin(),
+            op->get_pads_end(),
+            in_dilation);
+        return true;
+    }
+
+    template <element::Type_t ET>
+    bool evaluate(const shared_ptr<op::v1::ConvolutionBackpropData>& op,
+                  const HostTensorVector& outputs,
+                  const HostTensorVector& inputs)
+    {
+        const auto filter_data = inputs[1]->get_data_ptr<ET>();
+        auto out_data_ptr = outputs[0]->get_data_ptr<ET>();
+        const auto in_data_ptr = inputs[0]->get_data_ptr<ET>();
+        const auto& out_shape = outputs[0]->get_shape();
+        const auto& in_shape = inputs[0]->get_shape();
+        const auto& filter_shape = inputs[1]->get_shape();
+        Strides in_dilation(std::vector<size_t>(in_shape.size() - 2));
+        std::fill(in_dilation.begin(), in_dilation.end(), 1);
+        runtime::reference::convolution_backprop_in<typename element_type_traits<ET>::value_type>(
+            in_data_ptr,
+            filter_data,
+            out_data_ptr,
+            in_shape,
+            filter_shape,
+            out_shape,
+            in_dilation,
+            op->get_dilations(),
+            op->get_pads_begin(),
+            op->get_pads_end(),
+            op->get_strides());
+        return true;
+    }
+
+    template <element::Type_t ET>
+    bool evaluate(const shared_ptr<op::v0::CumSum>& op,
+                  const HostTensorVector& outputs,
+                  const HostTensorVector& inputs)
+    {
+        using T = typename element_type_traits<ET>::value_type;
+
+#define REF_CALL(U)                                                                                \
```

Comment:
Can this kind of macros be changed to a template instead? Those templates could produce an inline function that would be "pasted" by the compiler at the call site. This macro's argument is used as a type and thus using a template we get some level of type safety while a macro doesn't provide it.

Reply:
I've thought about it for a while and decided that I agree with you. Besides of mentioned types safety, templated functions much more convenient to edit than such defines.
So now this part is refactored.